[-Wunsafe-buffer-usage] Introduce std::array fixits

jkorous-apple opened the pull request.

jkorous-apple committed:
[-Wunsafe-buffer-usage] Move Strategy class to the header

It needs to be used from handleUnsafeVariableGroup function.

jkorous-apple committed:
[-Wunsafe-buffer-usage] Implement fixits from const size array to std::array

Array subscript on a const size array is not bounds-checked. The idiomatic
replacement is std::array which is bounds-safe in hardened mode of libc++.

This commit extends the fixit-producing machine to consider std::array as a
transformation target type and teaches it to handle the array subscript on const
size arrays with a trivial (empty) fixit.

The transformation is straightforward for most cases:
T array[N]; => std::array<T, N> array;
but for array of const pointers it needs to be:
T* const array[N]; => const std::array<T*, N> array;

This case and other special cases are to be implemented in follow-up
patches.

jkorous-apple committed:
[-Wunsafe-buffer-usage] Don't assert Array strategy is unimplemented

jkorous-apple committed:
[-Wunsafe-buffer-usage] Preserve spelling of array size in std::array fixits

Fixits related to arrays generally change type from T [N] to
std::array<T, N>. N has to be constant but doesn't have to be an integer
literal. It can be for example a constant.

The fixits need to preserve the spelling of the array size.
The assumption is that not specifying the numerical value directly was a
conscious decision of the original author and it'll very likely still apply to
the transformed code.

E. g. for
int buffer[SIZE];

if `buffer` is unsafe it should become:
std::array<int, SIZE> buffer;

rather than:
std::array<int, 42> buffer;

jkorous-apple committed:
[-Wunsafe-buffer-usage] Preserve array element type spelling in std::array fixits

In other words - don't use the type that the source code gets expanded
to in the AST as there's very likely a reason why it is represented via
a typedef or macro in the sourcode.

jkorous-apple committed:
[-Wunsafe-buffer-usage] Fix whitespace handling in array types

Specifically applies to text with array element type and array size for
constant size arrays. Trimming there strings on both sides and using
llvm::StringRef::trim().

jkorous-apple committed:
[-Wnsafe-buffer-usage][NFC] Fix and refactor tests

jkorous-apple committed:
[-Wunsafe-buffer-usage] Bail on fixits for arrays whose size is auto-deduced

C++ allows the array size to be omitted in declaration if it can be deduced from the
initializer.
E. g.:
int array [] = { 1, 2, 3 };

std::array<T, N> on the other hand requires the size N to be provided.

Currently the fixit produced would use empty space for N which would lead to a
compilation failure. This patch makes the fixit-producing machine just
bail in such cases and adds a FIXME to be addressed later.

jkorous-apple committed:
[-Wunsafe-buffer-usage][NFC] Add testcases

jkorous-apple committed:
[-Wunsafe-buffer-usage] Clang-format std::array fixits work

jkorous-apple committed:
[-Wunsafe-buffer-usage] Fix note message suggesting to use std::array

haoNoQ commented on clang/include/clang/Analysis/Analyses/UnsafeBufferUsage.h:
```diff
@@ -58,6 +95,8 @@ class UnsafeBufferUsageHandler {
 #endif
 
 public:
+  enum class TargetType { Span, Array };
```
Looks like a bit of dead code?

jkorous-apple replied:
Totally. Removed. Thanks!

haoNoQ commented on clang/lib/Analysis/UnsafeBufferUsage.cpp:
```diff
@@ -2495,10 +2470,113 @@ static FixItList fixVariableWithSpan(const VarDecl *VD,
+    //   T * foo[10]; => std::array<T *, 10> foo;
+    //
+    // However, for const pointers the transformation is different:
+    //   T * const foo[10]; => const std::array<T *, 10> foo;
```
Hmm, wouldn't `std::array<T *const, 10>` have the exact same behavior in practice? (Though I completely agree that `const std::array<T *, 10>` is more readable.) (Testbed: https://godbolt.org/z/aeohe6jW9)

jkorous-apple replied:
You're right, let me simplify this.

haoNoQ commented on clang/lib/Analysis/UnsafeBufferUsage.cpp:
```diff
@@ -2495,10 +2470,113 @@ static FixItList fixVariableWithSpan(const VarDecl *VD,
+                     Ctx.getLangOpts());
+    if (!MaybeElemTypeTxt)
+      return {};
+    const llvm::StringRef ElemTypeTxt = MaybeElemTypeTxt->trim();
```
Should `getRangeText()` trim automatically?

jkorous-apple replied:
For now I would leave it as is but if we later find that callers need to trim so often it gets annoying then let's reconsider.

haoNoQ commented on clang/test/SemaCXX/warn-unsafe-buffer-usage.cpp:
```diff
@@ -61,6 +61,7 @@ void testArraySubscripts(int *p, int **pp) {
       );
 
     int a[10];          // expected-warning{{'a' is an unsafe buffer that does not perform bounds checks}}
+                        // expected-note@-1{{change type of 'a' to 'std::array' to harden it}}
```
We'll probably have a broader discussion about wording but before I forget: "to harden it" may be misleading because we don't know/check whether library hardening is actually turned on.

jkorous-apple replied:
I am fine with broader discussion later but don't want to block the PR on it.

haoNoQ replied:
I think it's important to make sure it's not actively misleading, even if we plan to figure out the perfect phrasing later.  Maybe `to incapsulate it for hardening purposes`?

jkorous-apple replied:
I guess what we really want to say is "change the type of 'foo' to std::array so when you build the code with hardened libc++ all subscript operations on 'foo' will be automatically bounds-safe" but that is too long and too complex for this particular communication channel.

How about:
"to enable hardening"
"to get it hardened"
"to designate it for hardening"
"to register it for hardening"

jkorous-apple replied:
"so it can benefit from libc++ hardening"?

jkorous-apple committed:
[-Wunsafe-buffer-usage] Enable fixits for const ptr array declarations

jkorous-apple committed:
[-Wunsafe-buffer-usage] Remove unused TargetType enum

ziqingluo-90 commented on clang/test/SemaCXX/warn-unsafe-buffer-usage-fixits-local-var-array.cpp:
maybe we can have a test showing that multi-dimentional arrays will not be fixed?

jkorous-apple replied:
Excellent point!

jkorous-apple replied:
Added couple tests and an explicit check when we're creating fixit for unsafe array declaration.

ziqingluo-90 commented on clang/lib/Analysis/UnsafeBufferUsage.cpp:
```diff
@@ -2495,10 +2470,97 @@ static FixItList fixVariableWithSpan(const VarDecl *VD,
+    const SourceLocation LSqBracketLoc = NextTok->getLocation();
+
+    // Get the spelling of the array size as written in the source file
+    // (including macros, etc.).
```
I wish we could have a more elegant way to get the spelling of the array size.  I feel like `TypeLoc` is suppose to offer it but it does not behave as I expect in this case.  (`TypeLoc` also disappointed me in dealing with cv-qualifiers. `¯\_(ツ)_/¯`)

jkorous-apple replied:
I agree. I expect we'll get a better way of doing this and more when we adopt transformer library in the future. Getting the array size spelling might be already implemented or at least I expect to find better means of doing that than dealing with tokens.
For now I'm happy with a temporary solution while we bootstrap the rest of the machine.

jkorous-apple committed:
[-Wunsafe-buffer-usage] Explicitly unsupport fixits for multidimensional C arrays

Proper support for multidimensional arrays is outside of the scope of
initial std::array fixits.
This change makes sure that we don't accidentally emit fixits for multidimensional
arrays until that work is all done. It adds a specific check when transforming
VarDecl type to std::array. Also adds corresponding several testscases.

ziqingluo-90 commented on clang/lib/Analysis/UnsafeBufferUsage.cpp:
```diff
@@ -2495,10 +2471,100 @@ static FixItList fixVariableWithSpan(const VarDecl *VD,
+    const QualType &ArrayEltT = CAT->getElementType();
+    assert(!ArrayEltT.isNull() && "Trying to fix a non-array type variable!");
+    // FIXME: support multi-dimensional arrays
+    if (isa<clang::ConstantArrayType>(ArrayEltT))
```
is it possible that the element type is an array type but not a constant array type?

jkorous-apple replied:
You're right, for example it's possible to mix CSA with C variable length arrays.
https://clang.llvm.org/doxygen/classclang_1_1VariableArrayType.html

```
void foo(unsigned u) {
    int foo[5][u];
}

```
Let me fix this and add a testcase.

jkorous-apple replied:
...and on top of that the `if` should check for canonical type to not be fooled by typedefs & co.

ziqingluo-90 commented on clang/lib/Analysis/UnsafeBufferUsage.cpp:
```diff
@@ -2495,10 +2471,100 @@ static FixItList fixVariableWithSpan(const VarDecl *VD,
+      return {};
+    const llvm::StringRef ArraySizeTxt = MaybeArraySizeTxt->trim();
+    if (ArraySizeTxt.empty()) {
+      // FIXME: Support array size getting determined from the initializer.
```
In this case, maybe we can get the concrete size value from the array type directly since there is no spelling text for us to preserve?

jkorous-apple replied:
It is technically doable and the code would compile but if we want to preserve the "spirit of the code" then we can do a better job:

```
int arr1[] = {0, 1, 2};
// = fixit ===>
auto arr1 = std::to_array<int>({0, 1, 2});
```

The author of the original probably had a reason why they didn't hardcode the size in the first place.

ziqingluo-90 commented on clang/lib/Analysis/UnsafeBufferUsage.cpp:
```diff
@@ -2495,10 +2471,100 @@ static FixItList fixVariableWithSpan(const VarDecl *VD,
+       << IdentText->str();
+
+    FixIts.push_back(FixItHint::CreateReplacement(
+        SourceRange{D->getBeginLoc(), D->getTypeSpecEndLoc()}, OS.str()));
```
I'm wondering if it is possible to declare a constant array in a way that the type specifier is completely at the left-hand side of the identifier?   Like the form `T a;` while `T` represents a constant array type.   

How about this?  
```
int a[10];
decltype(a) b;
```

ziqingluo-90 replied:
and this maybe?
```
typedef int ArrayT[10];
ArrayT a;
```

jkorous-apple replied:
Excellent question!
It is indeed possible so I added a couple more testcases and verified that the implementation doesn't produce nonsense when encounters type sugar.

ziqingluo-90 reviewed (approved):
I left all my comments here.  I didn't think quite through for some of them so feel free to ignore if they don't make sense to you.  

Other parts LGTM!

jkorous-apple committed:
[-Wunsafe-buffer-usage] Avoid type sugar in array fixits

The code that creates fixits from C array to std::array assumes the array type is
not a sugared type like typedef as it would not produce correct fixits otherwise.
The patch adds a note about the condition under which the code works and testcases.
As a general precaution it also makes the format assumption explicit in iteration
over tokens of the declaration.

Contrary to that the higher layers need to use canonical (desugared) type in order
to decide which function is responsible for fixing an array VarDecl.

jkorous-apple committed:
[-Wunsafe-buffer-usage] Avoid multi-dimentional arrays properly

Array element can be for example VLA which given its non-constant length can't be transformed to std::array.

jkorous-apple commented on clang/lib/Analysis/UnsafeBufferUsage.cpp:
```diff
@@ -2495,10 +2471,100 @@ static FixItList fixVariableWithSpan(const VarDecl *VD,
+                                     UnsafeBufferUsageHandler &Handler) {
+  FixItList FixIts{};
+
+  if (auto CAT = dyn_cast<clang::ConstantArrayType>(D->getType())) {
```
Sorry for the confusion but I actually changed the conditions in the callers in the commit that addresses the type sugar.
https://github.com/llvm/llvm-project/pull/80084/commits/d49d5fd7e80463223aee9eac4ebd9e3c934cf9bd

My reasoning is that the layers above just check what kind of variable they are dealing with (e. g. pointer vs array) for which they need canonical type.
These rely on more specific functions to produce the fixits in a given context. Ultimately I see it as responsibility of `fixVarDeclWithArray` to handle any local variable of an array type and it shouldn't expect only canonical arrays. Maybe there will be some cases of type sugar for which it'll provide fixits in the future.

jkorous-apple committed:
[-Wunsafe-buffer-usage] Improve diagnostic note message

The hardening has two necessary conditions:
- using a bounds-safe standard library type in the source code
- compiling with hardened mode of libc++

The message shouldn't imply that just fixing the source code hardens it.
